### PR TITLE
Comments for entire codebase

### PR DESCRIPTION
**Pull request by ameller1994**

### CODE REVIEW

File: fast_movie.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
+import mdtraj as md
+import enspara
+from enspara import ra
+import os 
+import numpy as np
+from numpy import where
```
Comment (ameller1994) [inline fast_movie.py]: I would just do "np.where"

File: fast_movie.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+# Artur feedback
```
Comment (ameller1994) [inline fast_movie.py]: Would be nice to be able to call this from command line... look at enspara code for how to do that

File: fast_movie.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+folder = '/project/bowmanlab/ameller/fast-sims/ihm/FASTDistance_lad_Nterm_wt_5n69_min'
+out_folder = 'fast_movie_LA_10.xtc'
+
+# Setting up number of rounds, sims each round, file names, etc
+dirs =  next(os.walk(folder))[1]
```
Comment (ameller1994) [inline fast_movie.py]: Why this instead of a standard glob call? I think it's more readable with a glob call

File: fast_movie.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+sim_dirs = [e for e in dirs if 'gen' in e]
+total_rounds = len(sim_dirs)
+print(sim_dirs)
+gen0_path = os.path.join(folder,sim_dirs[0])
+sims_each_round = len(next(os.walk(gen0_path))[1])
+total_sims = total_rounds*sims_each_round
+dist_file = os.path.join(folder, 'msm/rankings' , 'distance_per_state'+ str(total_rounds-1)+'.npy') #must be customized
+start_state = np.flip(np.argsort(np.load(dist_file)))[0] #must be customzed
```
Comment (ameller1994) [inline fast_movie.py]: No need for np.flip

`start_state = np.argsort(np.load(dist_file))[-1]`

File: fast_movie.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+print(total_rounds,sims_each_round,start_state)
```
Comment (ameller1994) [inline fast_movie.py]: When you push this to enspara, your prints should become logging calls and have more informative messages about what is being printed

File: fast_movie.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+
+
+assigs = ra.load(os.path.join(folder, 'msm/data/assignments.h5'))
+pdb = md.load(os.path.join(folder, 'msm/prot_masses.pdb'))
+
+def find_new_start_state(prev_start_state, round_num):
+    # define window in assignments file for each round to find where state occurs
+    
+    window_start = total_sims - sims_each_round * (round_num+1)
```
Comment (ameller1994) [inline fast_movie.py]: So it seems like round_num must start from 0 here even though we usually start with the final round. I'd say this is a little confusing to someone else so I propose:

`window_start = sims_each_round * round_num`

File: fast_movie.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+start_state = np.flip(np.argsort(np.load(dist_file)))[0] #must be customzed
+print(total_rounds,sims_each_round,start_state)
+
+
+
+assigs = ra.load(os.path.join(folder, 'msm/data/assignments.h5'))
+pdb = md.load(os.path.join(folder, 'msm/prot_masses.pdb'))
+
+def find_new_start_state(prev_start_state, round_num):
+    # define window in assignments file for each round to find where state occurs
```
Comment (ameller1994) [inline fast_movie.py]: sims_each_round should be defined in this method

File: fast_movie.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    
+    window_start = total_sims - sims_each_round * (round_num+1)
+    window_end = total_sims - sims_each_round *(round_num)
+    traj, idx = np.where(assigs[window_start : window_end] == prev_start_state)
+    min_idx = np.argsort(idx)[0]
```
Comment (ameller1994) [inline fast_movie.py]: Is argsort right? Problems:

1) You don't need to sort the whole list since you only want the minimum
2) idx is already indices, right? So isn't the minimum index simply `np.min(idx)`? I found this confusing but kept reading your code and see that it's not a problem. Just a little hard to follow potentially for the reader.

File: fast_movie.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+pdb = md.load(os.path.join(folder, 'msm/prot_masses.pdb'))
+
+def find_new_start_state(prev_start_state, round_num):
+    # define window in assignments file for each round to find where state occurs
+    
+    window_start = total_sims - sims_each_round * (round_num+1)
+    window_end = total_sims - sims_each_round *(round_num)
+    traj, idx = np.where(assigs[window_start : window_end] == prev_start_state)
+    min_idx = np.argsort(idx)[0]   
+    traj, idx = [traj[min_idx], idx[min_idx]]
```
Comment (ameller1994) [inline fast_movie.py]:
```suggestion
    traj, idx = [traj[np.argmin(idx)], np.min(idx)]
```

File: fast_movie.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    new_start_state = assigs[window_start+traj][0]
+    print('traj: '+ str(window_start+traj) + ', idx: ' +str(idx) +', state: ' + str(new_start_state))
+    return(traj, idx, new_start_state)
```
Comment (ameller1994) [inline fast_movie.py]: May want to get ride of the tuple wrapping and just return the three variables separated by commas

File: fast_movie.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+def slice_traj(start_state, r):
```
Comment (ameller1994) [inline fast_movie.py]: r is not a very descriptive variable name

File: fast_movie.py [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+    # add to traj file
+    
+    traj, idx, start_state = find_new_start_state(start_state, r)
+    xtc_file = os.path.join(folder, 
+                             'msm/trajectories', 
+                             'trj_gen'+str(total_rounds-r-1).zfill(3)+'_kid'+str(traj).zfill(3)+'.xtc')
```
Comment (ameller1994) [inline fast_movie.py]: I would prefer if the round numbers corresponded to the actual FAST round numbers... that is r started at 9 instead of 0

File: fast_movie.py [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+    min_idx = np.argsort(idx)[0]   
+    traj, idx = [traj[min_idx], idx[min_idx]]
+    new_start_state = assigs[window_start+traj][0]
+    print('traj: '+ str(window_start+traj) + ', idx: ' +str(idx) +', state: ' + str(new_start_state))
+    return(traj, idx, new_start_state)
+
+def slice_traj(start_state, r):
+    # add to traj file
+    
+    traj, idx, start_state = find_new_start_state(start_state, r)
```
Comment (ameller1994) [inline fast_movie.py]: So this method returns a new start state right? Update the variable name so it's clear there's a difference between the new and old variables

File: fast_movie.py [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+    xtc_file = os.path.join(folder, 
+                             'msm/trajectories', 
+                             'trj_gen'+str(total_rounds-r-1).zfill(3)+'_kid'+str(traj).zfill(3)+'.xtc')
+    print(total_rounds-r-1)
+    tmp_traj = md.load(xtc_file, top=pdb) 
+    if idx!=0:
```
Comment (ameller1994) [inline fast_movie.py]: Is this necessary?

Can't you just use tmp_traj[0:idx] and if idx is 0 then that means an empty trajectory is returned and that whole round can be skipped (because it didn't get us any closer to our target)?

File: fast_movie.py [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+    else:
+        tmp_traj_slce = tmp_traj[0]
+
+    return(start_state, tmp_traj_slce)
+
+
+#output_xtc
+traj_list = {}
+key_list = []
+for r in range(total_rounds):
```
Comment (ameller1994) [inline fast_movie.py]: I would start from the highest round value in this loop because it will improve code readability

File: fast_movie.py [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+    #if r!= 4:
+    start_state, tmp_traj_slce = slice_traj(start_state, r)
+    traj_list[r] = tmp_traj_slce
+    key_list.append(r)
+
+traj_list_fin = []
```
Comment (ameller1994) [inline fast_movie.py]: Your use of data structures feels a little inefficient. I recommend using a single list where you store the output trajectories and then reversing that list. I don't think there's a need for a dictionary and there's almost never a need for a key list since you can always call a dictionary's .keys() variable.

File: fast_movie.py [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+[traj_list_fin.append(traj_list[e]) for e in np.flip(key_list)]
+        
+full_traj = md.join(traj_list_fin)
+
+#only if want to alter what is aligned when superposed
+#atoms=pdb.top.select('name CA and resi 20 to 700')
+full_traj.superpose(pdb)#, atom_indices=atoms, ref_atom_indices=atoms)
```
Comment (ameller1994) [inline fast_movie.py]: Consider having this as an option? I think it's ok for the molecule to drift in a FAST video.

File: fast_movie.py [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+
+#so output xtc isn't too large to load
+full_traj_sliced = full_traj[0:len(full_traj):10]
```
Comment (ameller1994) [inline fast_movie.py:96]: Also should probably be an option for the user to select

**Review by catherinekuhn17 — APPROVED** [review]
